### PR TITLE
Spock table param order validation

### PR DESCRIPTION
**Pull request by paplorinc**
Where blocks in Spock tests can contain tables with variable name headers.
To make them `@TypeChecked` or `@CompileStatic`, these params should be declared in the method parameter list also.

However, the Spock Where block rewriter (compiler) didn't check for the order of the params, causing hard to debug scenarios.

In this commit I added a check for parameter names' order, in case the where block only contained a table.

Example of an invalid test:

```Groovy
def 'invalid parameter order'(b, a) {
  expect: a != b
  where:  a | b
          1 | 2
}
```

which will give a compile error now!

### CODE REVIEW

File: spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy [hunk 1]
```diff
@@ -67,78 +65,93 @@ a
...
+    b == 'wow'
 
     where:
     a | b
-    3 | "wow"
+    3 | 'wow'
   }
 
-  def "pseudo-column can be omitted from parameter list"(a) {
+  def 'parameters must have same order as the table header'() {
```
Comment (paplorinc) [inline spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy]: this is the main test

File: spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy [hunk 2]
```diff
@@ -156,7 +169,18 @@ a | a
     3 | _
   }
 
-  def "tables can be mixed with other parameterizations"() {
+  def 'tables can be mixed with other parameterizations'() {
```
Comment (paplorinc) [inline spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy]: this is the negative test (i.e. validation ignored for the more complex case)

File: spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java [hunk 3]
```diff
@@ -215,10 +218,44 @@ private void rewriteTableLikeParameterization(ListIterator<Statement> stats) thr
       rows.add(row);
     }
 
+    validateParameterAndTableHeaderOrder(rows);
+
     for (List<Expression> column : transposeTable(rows))
       turnIntoSimpleParameterization(column);
   }
 
+  private void validateParameterAndTableHeaderOrder(List<List<Expression>> rows) throws InvalidSpecCompileException {
```
Comment (paplorinc) [inline spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java]: this is the validation itself

File: build.gradle [hunk 4]
```diff
@@ -16,7 +16,7 @@ description = "Spock Framework"
 
 ext {
   variants = [2.0, 2.3, 2.4]
-  variant = System.getProperty("variant") as BigDecimal ?: variants.first()
+  variant = System.getProperty("variant") as BigDecimal ?: variants.last()
```
Comment (leonard84) [inline build.gradle]: please don't include changes that are not directly related to your commit

File: gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.3-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.6-bin.zip
```
Comment (leonard84) [inline gradle/wrapper/gradle-wrapper.properties]: same here

File: spock-core/core.gradle [hunk 6]
```diff
@@ -25,13 +26,13 @@ jar {
     name = 'spock-core'
     instruction 'Export-Package', 'org.spockframework.*', 'spock.*'
     instruction 'Embed-Dependency', 'groovy;inline=false', 'junit;inline=false'
-    instruction 'Import-Package',
```
Comment (leonard84) [inline spock-core/core.gradle]: and here

File: spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java [hunk 7]
```diff
@@ -16,19 +16,22 @@
 
 package org.spockframework.compiler;
 
-import java.util.*;
-
 import org.codehaus.groovy.ast.*;
 import org.codehaus.groovy.ast.expr.*;
-import org.codehaus.groovy.ast.stmt.*;
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java]: Please fix your formatter to adhere to the given format -> keep the ordering and star imports

File: spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java [hunk 8]
```diff
@@ -66,11 +69,33 @@ private void rewrite() {
...
     handleFeatureParameters();
     createDataProcessorMethod();
   }
 
+  private void validateWhereBlockVariableOrder() {
+    try {
+      Parameter[] parameters = whereBlock.getParent().getAst().getParameters();
+      if (parameters.length == 0 || dataProcessorVars.size() == 0) return;
+
+      for (int i = 0; i < Math.min(parameters.length, dataProcessorVars.size()); i++)
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java]: `Math.min(parameters.length, dataProcessorVars.size())` should be saved into a local var

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy [hunk 9]
```diff
@@ -82,17 +83,19 @@ class EmbeddedSpecCompiler {
     doCompile "package apackage; $imports ${source.trim()}"
   }
 
-  Class compileSpecBody(String source) {
+  Class compileSpecBody(@Language(value = 'Groovy', prefix = 'class ASpec extends spock.lang.Specification { ', suffix = '}')
```
Comment (leonard84) [inline spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy]: the suffix is missing a newline to be identical to the inner code, same goes for every declaration

File: build.gradle (hunk 4 above)
Comment (paplorinc) [inline build.gradle]: I included it in a separate commit

File: spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java (hunk 7 above)
Comment (paplorinc) [inline spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java]: Left it at the default Idea setting, as there was no formatter committed to the project.
Will revert unrelated imports, if it bothers the reviewers.

File: spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java (hunk 8 above)
Comment (paplorinc) [inline spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java]: Will do

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy (hunk 9 above)
Comment (paplorinc) [inline spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy]: It doesn't need a new line, but I can add it, if it's confusing this way

File: spock-core/core.gradle (hunk 6 above)
Comment (paplorinc) [inline spock-core/core.gradle]: trailing whitespaces were trimmed by the IDE. I committed this separately.

**Comment by Fuud** [comment]
Why we use mapping by order? Mapping by name seems better.

**Comment by Fuud** [comment]
I want in near future let parameters in `where` depends on each other. The simples way to implement this - evaluate parameters left to right. Also it will be good to omit unnecessary parameters.

For example:
1) We use `testCase` parameter to specify test case id. It is used in rendering but it is not needed in test.
```
@Unroll("Test for test case #testCase")
def "test"(){
  expect:
    ...
  where:
    testCase | data1 | data2 ...
    "RLZ-126"|...|...
}
```

2) I want parameters to be evaluated depends on value of other parameters:
```
def "test transactions handling"(Transaction transaction){
  expect:
    ...
  where:
    changed_table | inserted_value | transaction
    "USERS" | new User(1, "Bob", ...) | createTransaction(changed_table, inserted_value)
}
```

Thus I prefer mapping by name and let developer to specify data in any order and  unnecessary parameters.

**Comment by paplorinc** [comment]
> 1) We use testCase parameter to specify test case id. It is used in rendering but it is not needed in test.

In your example you didn't have any params, but I presume you meant something like:
```Groovy
def 'test'(String data1, String data2) {
  where:
    testCase | data1 | data2 ...
```
right?

Also, in the second case, your request seems to be, that we should be able to avoid setting certain varible types.

I can change my commit to validate only existing parameters' order, i.e. the above code should be right, but
```Groovy
def 'test'(String data2, String data1) {
  where:
    testCase | data1 | data2 ...
```
should be invalid, because existing parameters have the wrong order.

Would that be ok :)?

File: spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy [hunk 10]
```diff
@@ -67,78 +63,96 @@ a
...
-    b == "wow"
+    b == 'wow'
 
     where:
     a | b
-    3 | "wow"
+    3 | 'wow'
   }
 
-  def "pseudo-column can be omitted from parameter list"(a) {
```
Comment (robfletcher) [inline spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy]: This kind of reformatting makes it very hard to see the wood for the trees in a pull request.

Comment (paplorinc) [inline spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy]: I tried separating it into commits for this exact reason. What do you suggest I do instead?

**Comment by leonard84** [comment]
Please squash (e.g. rebase interactive with squash) all into one commit and **exclude** all reformatting of existing code, also exclude the changes to the gradle, groovy version, and default variant you can make a *separate* PR for that. 

Please motivate the inclusion of IntelliJ Annotations.

If there is a problem with `@PendingFeature` please file a bug report with an [SSCCE](http://sscce.org).

**Comment by paplorinc** [comment]
sure, will do

**Comment by leonard84** [comment]
@paplorinc you wouldn't need to close this PR, a force push to the branch should have worked.

**Comment by paplorinc** [comment]
Yes, but this branch started with a lowercase letter and the rest of the PRs were capitalized :p
